### ISSUE TITLE
handling complex dtypes

### ISSUE THREAD

**Issue by grlee77** [PyWavelets/pywt#96]
This is a request for feedback on a potential enhancement to handle complex data for transforms provided by pywt.  Personally, I work with MRI images where the data is complex and both the image magnitude and phase are often of interest.

Currently one can perform complex transforms via separate calls on the real and imaginary parts, but this is a bit cumbersome:
```python
coeffs_real = pywt.dwt(data.real, wavelet)
coeffs_imag = pywt.dwt(data.imag, wavelet)

reconstructed_data = pywt.idwt(coeffs_real, wavelet) + 1j*pywt.idwt(coeffs_imag, wavelet)
```

Is this too niche a case to be worth supporting here in pywt?  I see three potential approaches in ascending order of the work required, but would be happy for feedback on others:

1. leave it up to the user to handle as above.
2. modify functions such as dwt to accept complex inputs, but run the real and complex parts   independently using the existing C codebase.  This would essentially operate like in the example above, but the data/coeffs input/returned would be complex.  The separate real/imag calls would be transparent to users of the functions.
3. update the C code to support ``np.complex64`` and ``np.complex128`` in addition to the currently supported ``np.float32`` and ``np.float64``.

Even if it is determined not to add complex support, I think we should raise errors on complex input as the current code will just discard the imaginary part.  This does issue a warning ``ComplexWarning: Casting complex values to real discards the imaginary part``, but that could be easy to miss.

**Comment by kwohlfahrt** [comment]
I like this idea. I'd be happy to implement option 3, but suspect it might bite us on MSVC since it's a C99 thing. If you can compile something like the snippet below on MSVC 2008 let me know and I'll get cracking.

```
#include <complex.h>

int main(void){
    float complex i = {1.0, 2.0};
    float j = 3.0;
    float complex k = i * j;
    return 0;
}
```

Otherwise I think option 2 is the way to go. I think it'll be reasonably quick to do on top of the n-dimensional bits.

**Comment by grlee77** [comment]
I don't have access to MSVC2008 at the moment, but I'm pretty certain that won't compile there.  

numpy defines their own C89 compatible complex types such as the following:

```c
#if defined(NPY_USE_C99_COMPLEX) && defined(NPY_HAVE_COMPLEX_FLOAT)
typedef complex float npy_cfloat;
#else
typedef struct { float real, imag; } npy_cfloat;
#endif
```

**Comment by grlee77** [comment]
actually, the code I pasted above is commented out in numpy and it always uses
```c
typedef struct { double real, imag; } npy_cdouble;
typedef struct { float real, imag; } npy_cfloat;
```

**Comment by kwohlfahrt** [comment]
Ah. I would go for option 2 then. As I understand it, adding this using C99 data types is simple, since we can just add another type to the .src expansion and it should just work. Adding it with a special struct data type is not, because there isn't a multiplication operator defined for them, so we'd have to duplicate all of the functions and add a `output[o].real = filter[j] * input[i].real; output[o].imag = filter[j] * input[i].imag` in place of `output[o] = filter[j] * input[i];` or something. It *could* probably be done with a macro, but I'm not particularly keen to go there unless there's a big performance argument.

**Comment by grlee77** [comment]
apparently there are still issues with C99 complex even in recent MSVC compilers, so I guess option 3 is out for now. e.g.:
https://github.com/numpy/numpy/pull/6171
https://github.com/numpy/numpy/issues/4896

a version of option 2 above is in PR #99

**Comment by grlee77** [comment]
This was resolved by #99
